pkg/core/chain/synchronizer.go

```diff
@@ -119,19 +122,31 @@ func (s *Synchronizer) ProcessBlock(m message.Message) (res []bytes.Buffer, err
 		return
 	}
 
-	if blk.Header.Height > s.highestSeen {
-		s.highestSeen = blk.Header.Height
+	s.lock.Lock()
+	currState := s.state
+	if blk.Header.Height > s.highestSeenHeight {
+		s.highestSeenHeight = blk.Header.Height
 	}
+	s.lock.Unlock()
+
+	var newState syncState
+	newState, res, err = currState(currentHeight, blk)
+
+	s.lock.Lock()
+	s.state = newState
+	s.lock.Unlock()
```

Comment:
Nit: this might look a bit nicer if extrapolated into its own function.

```golang
func (s *Synchronizer) setState(state syncState) {
    s.lock.Lock()
    s.state = state
    s.lock.Unlock()
}
```